packages/amplify-opensearch-simulator/index.js

```diff
@@ -0,0 +1,293 @@
+  }
+
+  terminate() {
+    // already exited
+    if (this.proc.exitCode != null) {
+      return this.proc.exitCode;
+    }
+    this.proc.kill();
+    return fromEvent(this.proc, 'exit');
+  }
+}
+
+const wait = ms => {
+  let timeoutHandle;
+  const promise = new Promise(accept => {
+    timeoutHandle = setTimeout(accept, ms);
+  });
+
+  return {
+    promise,
+    cancel: () => {
+      clearTimeout(timeoutHandle);
+    },
+  };
+};
+
+function buildArgs(options) {
+  const args = [];
+
+  if (options.clusterName) {
+    args.push(`-Ecluster.name=${options.clusterName}`);
+  }
+
+  if (options.nodeName) {
+    args.push(`-Enode.name=${options.nodeName}`);
+  }
+
+  if (options.port) {
+    args.push(`-Ehttp.port=${options.port}`);
+  }
+
+  if (options.type) {
+    args.push(`-Ediscovery.type=${options.type}`);
+  }
+
+  return args;
+}
+
+async function launch(pathToOpenSearchLocal, givenOptions = {}, retry = 0, startTime = Date.now()) { 
+  if (process.platform.startsWith('win')) {
+    throw new Error('Cannot launch OpenSearch simulator on windows OS');
+  }
+
+  log.info('launching OpenSearch Simulator with options: ', { retry, givenOptions });
+  // launch will retry but ensure it will not retry indefinitely.
+  if (retry >= maxRetries) {
+    throw new Error('Max retries hit for starting OpenSearch simulator');
+  }
+
+  try {
+    await ensureOpenSearchLocalExists(pathToOpenSearchLocal);
+  } catch (error) {
+    throw new Error('Failed to setup local OpenSearch simulator');
+  }
+
+  let { port } = givenOptions;
+  if (!port) {
+    port = await detectPort(basePort);
+    log.info('found open port', { port });
+  } else {
+    const freePort = await detectPort(port);
+    if (freePort !== port) {
+      throw new Error(`Port ${port} is not free. Please use a different port`);
+    }
+  }
+  const opts = { ...defaultOptions, ...givenOptions, port };
+
+  const args = buildArgs(opts);
+  log.info('Spawning OpenSearch Simulator:', { args, cwd: pathToOpenSearchLocal });
+  const openSearchBinPath = await getPathToOpenSearchBinary();
+
+  const proc = execa(openSearchBinPath, args, {
+    cwd: pathToOpenSearchLocal,
+  });
+
+  function startingTimeout() {
+    log.error('Failed to start within timeout');
+    // ensure process is halted.
+    proc.kill();
+    const err = new Error('start has timed out!');
+    err.code = 'timeout';
+    throw err;
+  }
+
+  // define this now so we can use it later to remove a listener.
+  let prematureExit;
+  let waiter;
+  /*
+   This is a fairly complex set of logic, similar to the DynamoDB emulator, 
+   to retry starting the emulator if it fails to start. We need this logic due
+   to possible race conditions between when we find an open
+   port and bind to it. This situation is particularly common
+   in jest where each test file is running in it's own process
+   each competing for the open port.
+  */
+  try {
+    waiter = wait(opts.startTimeout);
```

Comment:
can we break this into a few functions (maybe one for each promise that is racing) so it's easier to read?